core/trino-parser/src/main/java/io/trino/sql/jsonpath/PathTreeBuilder.java

```diff
@@ -0,0 +1,419 @@
+import io.trino.sql.jsonpath.tree.Filter;
+import io.trino.sql.jsonpath.tree.FloorMethod;
+import io.trino.sql.jsonpath.tree.IsUnknownPredicate;
+import io.trino.sql.jsonpath.tree.JsonNullLiteral;
+import io.trino.sql.jsonpath.tree.JsonPath;
+import io.trino.sql.jsonpath.tree.KeyValueMethod;
+import io.trino.sql.jsonpath.tree.LastIndexVariable;
+import io.trino.sql.jsonpath.tree.LikeRegexPredicate;
+import io.trino.sql.jsonpath.tree.MemberAccessor;
+import io.trino.sql.jsonpath.tree.NamedVariable;
+import io.trino.sql.jsonpath.tree.NegationPredicate;
+import io.trino.sql.jsonpath.tree.PathNode;
+import io.trino.sql.jsonpath.tree.Predicate;
+import io.trino.sql.jsonpath.tree.PredicateCurrentItemVariable;
+import io.trino.sql.jsonpath.tree.SizeMethod;
+import io.trino.sql.jsonpath.tree.SqlValueLiteral;
+import io.trino.sql.jsonpath.tree.StartsWithPredicate;
+import io.trino.sql.jsonpath.tree.TypeMethod;
+import io.trino.sql.tree.BooleanLiteral;
+import io.trino.sql.tree.DecimalLiteral;
+import io.trino.sql.tree.DoubleLiteral;
+import io.trino.sql.tree.LongLiteral;
+import io.trino.sql.tree.StringLiteral;
+import org.antlr.v4.runtime.tree.TerminalNode;
+
+import java.util.Optional;
+
+public class PathTreeBuilder
+        extends JsonPathBaseVisitor<PathNode>
+{
+    @Override
+    public PathNode visitPath(JsonPathParser.PathContext context)
+    {
+        boolean lax = context.pathMode().LAX() != null;
+        PathNode path = visit(context.pathExpression());
+        return new JsonPath(lax, path);
+    }
+
+    @Override
+    public PathNode visitDecimalLiteral(JsonPathParser.DecimalLiteralContext context)
+    {
+        return new SqlValueLiteral(new DecimalLiteral(context.getText()));
+    }
+
+    @Override
+    public PathNode visitDoubleLiteral(JsonPathParser.DoubleLiteralContext context)
+    {
+        return new SqlValueLiteral(new DoubleLiteral(context.getText()));
+    }
+
+    @Override
+    public PathNode visitIntegerLiteral(JsonPathParser.IntegerLiteralContext context)
+    {
+        return new SqlValueLiteral(new LongLiteral(context.getText()));
+    }
+
+    @Override
+    public PathNode visitStringLiteral(JsonPathParser.StringLiteralContext context)
+    {
+        return new SqlValueLiteral(new StringLiteral(unquote(context.STRING().getText())));
+    }
+
+    private static String unquote(String quoted)
+    {
+        return quoted.substring(1, quoted.length() - 1)
+                .replace("\"\"", "\"");
+    }
+
+    @Override
+    public PathNode visitNullLiteral(JsonPathParser.NullLiteralContext context)
+    {
+        return new JsonNullLiteral();
+    }
+
+    @Override
+    public PathNode visitBooleanLiteral(JsonPathParser.BooleanLiteralContext context)
+    {
+        return new SqlValueLiteral(new BooleanLiteral(context.getText()));
+    }
+
+    @Override
+    public PathNode visitContextVariable(JsonPathParser.ContextVariableContext context)
+    {
+        return new ContextVariable();
```

Comment:
nit: add comment
``` 
// drop leading `$`
```

Reply:
Actually, `$` is not a marker that you need to drop from the beginning of the path, though most of the use cases will look like this. `$` is a variable that can be used throughout the path, just like the named variables `$name`.

Reply:
Did I position comment correclty?
It was meant to be at `return new NamedVariable(namedVariable.getText().substring(1))` in `namedVariable`.

And it the `1` is about cutting `$` out, right?

Reply:
Yeah, it was another method. 
Added the comment.